File: src/test/java/fi/eonwe/wikilinks/RouteFinderTest.java
package fi.eonwe.wikilinks;

import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Set;
import java.util.stream.Stream;

import com.google.common.collect.Maps;
import fi.eonwe.wikilinks.leanpages.OrderedPage;
import org.jgrapht.EdgeFactory;
import org.jgrapht.VertexFactory;
import org.jgrapht.alg.DijkstraShortestPath;
import org.jgrapht.generate.RandomGraphGenerator;
import org.jgrapht.graph.SimpleDirectedGraph;
import org.junit.Ignore;
import org.junit.Test;
import static org.junit.Assert.assertEquals;

/**
 */
public class RouteFinderTest {

    private static class IntEdge {
        public final int start;
        public final int end;

        private IntEdge(Integer start, Integer end) {
            this.start = start;
            this.end = end;
        }

        @Override
        public String toString() {
            return String.format("%d->%d", start, end);
        }
    }

    private static class VF implements VertexFactory<Integer> {

        private int vertexCount = 0;

        @Override
        public Integer createVertex() {
            return vertexCount++;
        }

        public void reset() { vertexCount = 0; }
    }

    private static class EF implements EdgeFactory<Integer, IntEdge> {
        @Override
        public IntEdge createEdge(Integer sourceVertex, Integer targetVertex) {
            return new IntEdge(sourceVertex, targetVertex);
        }
    }

    @Test
    @Ignore("Too slow")
    public void findsTheSameSizedRoutes() {
        final Random rng = new Random(0xcafebabe);
        final int vertexCount = 1000;
        final int edgeCount = vertexCount * Math.min(vertexCount / 2, 100);
        RandomGraphGenerator<Integer, IntEdge> generator = new RandomGraphGenerator<>(vertexCount, edgeCount);
        final int repeats = 10;
        final int innerRepeats = 10;
        for (int i = 0; i < repeats; i++) {
            SimpleDirectedGraph<Integer, IntEdge> graph = new SimpleDirectedGraph<>(new EF());
            Map<String, Integer> resultMap = Maps.newHashMap();
            generator.generateGraph(graph, new VF(), resultMap);
            Map<Integer, OrderedPage> orderedPageMap = createFromGraph(graph);
            Set<Integer> vertices = graph.vertexSet();
            for (int j = 0; j < innerRepeats; j++) {
                int startVertex, endVertex;
                do {
                    startVertex = rng.nextInt(vertices.size());
                    endVertex = rng.nextInt(vertices.size());
                } while (startVertex == endVertex);
                long dijkstraStart = System.currentTimeMillis();
                DijkstraShortestPath<Integer, IntEdge> dijkstra = new DijkstraShortestPath<>(graph, startVertex, endVertex);
                int[] route = toIntArray(dijkstra.getPathEdgeList());
                long jgraphtTime = System.currentTimeMillis() - dijkstraStart;

                WikiRoutes.PageMapper mapper = fromMap(orderedPageMap);
                long myStart = System.currentTimeMillis();
                int[] myRoute = RouteFinder.find(startVertex, endVertex, mapper, null);
                long myTime = System.currentTimeMillis() - myStart;
//                System.out.printf("%d vertices, %d edges: JGraphT=%d ms, RouteFinder=%d ms%n", vertexCount, (long) edgeCount * vertexCount, jgraphtTime, myTime);
                assertEquals(route.length, myRoute.length);
            }
        }
    }

    private static int[] toIntArray(List<IntEdge> path) {
        if (path.isEmpty()) return new int[0];
        int[] startPoints = Stream.concat(path.stream().map(intEdge -> intEdge.start), Stream.of(path.get(path.size() - 1).end)).mapToInt(Number::intValue).toArray();
        return startPoints;
    }

    private static WikiRoutes.PageMapper fromMap(Map<Integer, OrderedPage> map) {
        return (pageIndex, c) -> map.get(pageIndex).forEachLinkIndex(c);
    }

    private static Map<Integer, OrderedPage> createFromGraph(SimpleDirectedGraph<Integer, IntEdge> graph) {
        Map<Integer, OrderedPage> result = Maps.newHashMap();
        for (Integer vertex : graph.vertexSet()) {
            Set<IntEdge> out = graph.outgoingEdgesOf(vertex);
            int[] targets = out.stream().mapToInt(e -> e.end).toArray();
            OrderedPage page = new OrderedPage(null, targets) {
                private final int id = vertex;
                @Override
                public int getId() {
                    return id;
                }
            };
            result.put(vertex, page);
        }
        return result;
    }


}

File: src/test/kotlin/fi/eonwe/wikilinks/RouteFinderTest.kt
package fi.eonwe.wikilinks

import fi.eonwe.wikilinks.leanpages.BufferWikiPage
import fi.eonwe.wikilinks.leanpages.OrderedPage
import io.kotest.matchers.shouldBe
import org.jgrapht.EdgeFactory
import org.jgrapht.VertexFactory
import org.jgrapht.alg.DijkstraShortestPath
import org.jgrapht.generate.RandomGraphGenerator
import org.jgrapht.graph.SimpleDirectedGraph
import org.junit.jupiter.api.Test
import java.util.Random
import java.util.function.IntConsumer
import java.util.stream.Stream
import kotlin.math.min

/**
 */
class RouteFinderTest {
    private class IntEdge(val start: Int, val end: Int) {
        override fun toString(): String {
            return String.format("%d->%d", start, end)
        }
    }

    private class VF : VertexFactory<Int?> {
        private var vertexCount = 0

        override fun createVertex(): Int {
            return vertexCount++
        }

    }

    private class EF : EdgeFactory<Int, IntEdge> {
        override fun createEdge(sourceVertex: Int, targetVertex: Int): IntEdge {
            return IntEdge(sourceVertex, targetVertex)
        }
    }

    @Test
    fun findsTheSameSizedRoutes() {
        val rng = Random(-0x35014542)
        val vertexCount = 1000
        val edgeCount = vertexCount * min(vertexCount / 2, 100)
        val generator: RandomGraphGenerator<Int?, IntEdge?> =
            RandomGraphGenerator<Int?, IntEdge?>(vertexCount, edgeCount)
        val repeats = 10
        val innerRepeats = 10
        repeat(repeats) {
            val graph = SimpleDirectedGraph(EF())
            val resultMap = mutableMapOf<String?, Int?>()
            generator.generateGraph(graph, VF(), resultMap)
            val orderedPageMap: MutableMap<Int, OrderedPage> = createFromGraph(graph)
            val vertices = graph.vertexSet()
            repeat(innerRepeats) {
                var startVertex: Int
                var endVertex: Int
                do {
                    startVertex = rng.nextInt(vertices.size)
                    endVertex = rng.nextInt(vertices.size)
                } while (startVertex == endVertex)
                val dijkstra: DijkstraShortestPath<Int?, IntEdge?> = DijkstraShortestPath(graph, startVertex, endVertex)
                val route: IntArray = toIntArray(dijkstra.getPathEdgeList().mapNotNull { it })

                val mapper: WikiRoutes.PageMapper = fromMap(orderedPageMap)
                val myRoute = RouteFinder.find(startVertex, endVertex, mapper, null)

                route.size shouldBe myRoute.size
            }
        }
    }

    companion object {
        private fun toIntArray(path: List<IntEdge>): IntArray {
            if (path.isEmpty()) return IntArray(0)
            val startPoints =
                Stream.concat(
                    path.stream().map { intEdge: IntEdge -> intEdge.start }, Stream.of(
                        path[path.size - 1].end
                    )
                ).mapToInt { obj: Int? -> obj!! }.toArray()
            return startPoints
        }

        private fun fromMap(map: MutableMap<Int, OrderedPage>): WikiRoutes.PageMapper {
            return object : WikiRoutes.PageMapper {
                override fun forEachLinkIndex(pageIndex: Int, c: IntConsumer) {
                    return map[pageIndex]!!.forEachLinkIndex(c)
                }
            }
        }

        private fun createFromGraph(graph: SimpleDirectedGraph<Int, IntEdge>): MutableMap<Int, OrderedPage> {
            val result: MutableMap<Int, OrderedPage> = mutableMapOf()
            for (vertex in graph.vertexSet()) {
                val out = graph.outgoingEdgesOf(vertex)
                val targets = out.stream().mapToInt { e: IntEdge? -> e!!.end }.toArray()
                val bwp = BufferWikiPage.createFrom(vertex, IntArray(0), "", false)
                val page = OrderedPage(bwp, targets)
                result[vertex] = page
            }
            return result
        }
    }
}